takashi-task-framework.md

```diff
@@ -0,0 +1,265 @@
+
+## Design
+
+Assuming that we're trying to design a task framework that allows for
+our users to run arbitrary things:
+
+Let's start out with something like systemd:
+
+First, we need some way to define a task: this part needs to be
+sufficiently expressive for the end user.
+
+It's important that our task runner allows for dependencies (like check
+if lib-x is installed) and that it can check for the correct functionality
+(lib-x must be able to do y). This part will be a lot like autotools,
+where there are checks that occur before the task is run to make sure
+the task runner is in a state where it can run the proper tasks.
+
+One way to make this light without adding a lot of undue burden onto the
+task running implementation is to allow shebang scripts for these
+dependency checks.
+
+Let's say we want to make sure that `pandoc` is on the `$PATH`, because
+a task we want to describe requires pandoc.
+
+// check-pandoc.sh
+
+```sh
+#!/usr/bin/env bash
+
+which pandoc # returns a status code of 0 if found, 1 if not.
+```
+
+We can see that this command will return a `0` if `pandoc` is on the
+`$PATH`, or a `1` if not.
+
+Thus, we can compose our dependencies by having our task description run
+bash scripts in the order they are placed in the config file, until they
+all return 0 or one script returns a non-zero exit code. In case a
+script returns a non-zero exit code, we can bubble up the error to the
+user (clearly this task can't run).
+
+Let's assume we'll use some key-value format, like an INI format.
+
+```ini
+[Checks]
+check-pandoc="./check-pandoc"
+```
+
+Next, we want to make sure that we can pass environment variables, and
+command line arguments to our script:
+
+```ini
+[Checks]
+check-pandoc="./check-pandoc"
+
+[Environment]
+DOCKER_PORT="0.0.0.0"
+
+[Args]
+args="site.md -o site.html --template=_template.html"
+
+[Command]
+command="pandoc"
+```
+
+Finally, we need to find some way to express dependencies:
+
+```ini
+[Dependencies]
+Requires=libpng.service # make sure that libpng is running.
+After=network.target # make sure the network is running
+
+[Service]
+User=test
+Group=test
+WorkingDirectory=/tmp/pandoc
+Restart=never
+LimitMem=infinity # infinite memory
+MaxTime=2s # if it doesn't return in 2s, kill this job
+
+[Checks]
+check-pandoc="./check-pandoc"
+
+[Environment]
+DOCKER_PORT="0.0.0.0"
+
+[Args]
+args="site.md -o site.html --template=_template.html"
+
+[Command]
+command="pandoc"
+```
+
+Where libpng.service might look like this:
+
+```ini
+[Dependencies]
+Requires=libjpeg.service # make sure that libjpeg is running
+```
+
+So we can express dependencies of dependencies. We'll resolve
+dependencies of dependencies by doing a topological sort, where we start
+at the top of the dependency list and make sure they all successfully
+run before going to the task at hand.
+
```

Comment:
you talk about writing a topological sort. I love the idea and think it makes a lot of sense here but what would be considered the in-degree of a dependency? how do we build our graph beforehand? do we expand all the dependencies into some sort of tree-like structure? how can we ensure that we aren't reverifying common dependencies of major projects like log4j or commonjs as this will slow down our process? 


Is there some mechanism that we can use to make sure that after we verify a specific dependency version we don't have to put in the work of reverifying those dependencies? This could become a bottleneck in a system that executes 100,000 tasks all with many dependencies.

Reply:
I see a few questions here:

1. What would be the in-degree of a dependency?

- We'd first have to traverse all of the dependencies and be able to create a graph of dependencies. If the graph is a DAG, which we can check for, there is a viable topological sort.
We can run some logic to check for that, and run a compile time error if there is no DAG.
Most languages have a topological sort library (ruby, python in the stdlib) and even the coreutils has tsort so we can offload that to a library if we'd like.

2. How can we ensure that we aren't reverifying common dependencies of major projects like log4j or commonjs as this will slow down our process?

- There's another question here: Should there be only one of a dependency (like java, c, c++) or can there be multiple dependencies at different versions (like ruby, js, rust) allow? I'll go with the latter since I'm prefer it for not breaking builds.
- To ensure that we're not reverifying common dependencies, we could use a set to keep track of a tuple of (dependency, version) to deduplicate versions. When first generating the graph for the toplogical sort, we can deduplicate it by putting every possible node into the set and then trying to generate a topological sort from the remaining nodes.